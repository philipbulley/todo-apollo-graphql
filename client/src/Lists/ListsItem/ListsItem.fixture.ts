import { ListsItemProps } from './ListsItem';

export const SHOPPING_LIST: ListsItemProps['list'] = {
  id: 'abc',
  name: 'Shopping',
  createdAt: '2020-02-02 20:00:00',
  updatedAt: '2020-02-03 12:30:40'
};
