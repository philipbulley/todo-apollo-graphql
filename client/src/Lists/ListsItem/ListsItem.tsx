import React, { FunctionComponent, useCallback, useMemo } from 'react';
import CardContent from '@material-ui/core/CardContent';
import Typography from '@material-ui/core/Typography';
import Card from '@material-ui/core/Card';
import { TodoList } from '../../__generated__/graphql';
import { Optional } from 'utility-types';
import { formatDistance } from 'date-fns';
import ScheduleIcon from '@material-ui/icons/Schedule';
import styled from 'styled-components/macro';
import ButtonBase from '@material-ui/core/ButtonBase';
import { useHistory } from 'react-router-dom';

export type ListsItemProps = {
  list: Optional<TodoList, 'items'>;
};

const ListsItem: FunctionComponent<ListsItemProps> = ({ list }) => {
  const { push } = useHistory();
  const updatedAt = useMemo(
    () =>
      formatDistance(new Date(list.updatedAt), new Date().getTime(), {
        addSuffix: true
      }),
    [list.updatedAt]
  );
  const handleClick = useCallback(() => push(`/lists/${list.id}`), [
    push,
    list.id
  ]);

  return (
    <CardButtonBase onClick={handleClick}>
      <Card>
        <CardContent>
          <Name variant="h2">{list.name}</Name>
          <ScheduleIconStyled fontSize="small" />
          <Typography variant="caption">{updatedAt}</Typography>
        </CardContent>
      </Card>
    </CardButtonBase>
  );
};

export default ListsItem;

const ScheduleIconStyled = styled(ScheduleIcon)`
  margin-right: 4px;
  vertical-align: middle;
`;

const Name = styled(Typography).attrs({ variant: 'h2' })`
  margin-bottom: 40px;
`;

const CardButtonBase = styled(ButtonBase)`
  display: block;
  width: 100%;
  text-align: left;
  cursor: pointer;
`;
