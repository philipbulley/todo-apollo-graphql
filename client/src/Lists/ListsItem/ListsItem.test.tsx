import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { SHOPPING_LIST } from './ListsItem.fixture';
import ListsItem from './ListsItem';
import { advanceTo, clear } from 'jest-date-mock';
import { useHistory } from 'react-router-dom';

// Mock icon so we don't pollute snapshot with its SVG implementation details
jest.mock('@material-ui/icons/Schedule', () => () => (
  <i data-testid="ScheduleIcon" />
));

jest.mock('react-router-dom', () => ({
  useHistory: jest.fn().mockReturnValue({
    push: jest.fn()
  })
}));

describe('ListItem', () => {
  beforeAll(() => {
    advanceTo('2020-02-14T18:28:34.231Z');
  });

  afterAll(() => {
    clear();
  });

  it('should render', () => {
    const { container } = render(<ListsItem list={SHOPPING_LIST} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should navigate to a specific todo list', () => {
    const { container } = render(<ListsItem list={SHOPPING_LIST} />);

    const { push } = useHistory();

    if (!container.firstElementChild) {
      throw new Error(`Can't find firstElementChild`);
    }

    fireEvent.click(container.firstElementChild);

    expect(push).toHaveBeenCalledTimes(1);
    expect(push).toHaveBeenCalledWith(`/lists/${SHOPPING_LIST.id}`);
  });
});
