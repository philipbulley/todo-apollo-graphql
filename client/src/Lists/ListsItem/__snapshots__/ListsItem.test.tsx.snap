// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem should render 1`] = `
.c1 {
  margin-bottom: 40px;
}

.c0 {
  display: block;
  width: 100%;
  text-align: left;
  cursor: pointer;
}

<button
  class="MuiButtonBase-root c0"
  tabindex="0"
  type="button"
>
  <div
    class="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      class="MuiCardContent-root"
    >
      <h2
        class="MuiTypography-root c1 MuiTypography-h2"
      >
        Shopping
      </h2>
      <i
        data-testid="ScheduleIcon"
      />
      <span
        class="MuiTypography-root MuiTypography-caption"
      >
        11 days ago
      </span>
    </div>
  </div>
  <span
    class="MuiTouchRipple-root"
  />
</button>
`;
